shared_model/validators/commands_validator.hpp

```diff
@@ -0,0 +1,303 @@
+
+          return res;
+        }
+
+        ReasonsGroupType operator()(
+            const detail::PolymorphicWrapper<interface::CreateAsset> &ca)
+            const {
+          std::string class_name = "CreateAccount";
+          ReasonsGroupType res;
+
+          validateAssetName(res, ca->assetName());
+          validateDomainId(res, ca->domainId());
+          validatePrecision(res, ca->precision());
+
+          return res;
+        }
+
+        ReasonsGroupType operator()(
+            const detail::PolymorphicWrapper<interface::CreateDomain> &cd)
+            const {
+          std::string class_name = "CreateDomain";
+          ReasonsGroupType res;
+
+          validateDomainId(res, cd->domainId());
+
+          return res;
+        }
+
+        ReasonsGroupType operator()(
+            const detail::PolymorphicWrapper<interface::CreateRole> &cr) const {
+          std::string class_name = "CreateRole";
+          ReasonsGroupType res;
+
+          validateRoleId(res, cr->roleName());
+
+          return res;
+        }
+
+        ReasonsGroupType operator()(
+            const detail::PolymorphicWrapper<interface::GrantPermission> &gp)
+            const {
+          std::string class_name = "GrantPermission";
+          ReasonsGroupType res;
+
+          validateAccountId(res, gp->accountId());
+
+          return res;
+        }
+
+        ReasonsGroupType operator()(
+            const detail::PolymorphicWrapper<interface::RemoveSignatory> &rs)
+            const {
+          std::string class_name = "RemoveSignatory";
+          ReasonsGroupType res;
+
+          validateAccountId(res, rs->accountId());
+          validatePubkey(res, rs->pubkey());
+
+          return res;
+        }
+        ReasonsGroupType operator()(
+            const detail::PolymorphicWrapper<interface::RevokePermission> &rp)
+            const {
+          std::string class_name = "RevokePermission";
+          ReasonsGroupType res;
+
+          validateAccountId(res, rp->accountId());
+          validatePermission(res, rp->permissionName());
+
+          return res;
+        }
+        ReasonsGroupType operator()(
+            const detail::PolymorphicWrapper<interface::SetQuorum> &sq) const {
+          std::string class_name = "SetQuorum";
+          ReasonsGroupType res;
+
+          validateAccountId(res, sq->accountId());
+          validateQuorum(res, sq->newQuorum());
+
+          return res;
+        }
+
+        ReasonsGroupType operator()(
+            const detail::PolymorphicWrapper<interface::TransferAsset> &ta)
+            const {
+          std::string class_name = "SetQuorum";
+          ReasonsGroupType res;
+
+          validateAccountId(res, ta->srcAccountId());
+          validateAccountId(res, ta->destAccountId());
+          validateAssetId(res, ta->assetId());
+          validateAmount(res, ta->amount());
+
+          return res;
+        }
+
+       private:
+        void validateAccountId(
```

Comment:
These bunch of methods are quite similar, maybe create generic method and use aliases for them?

Reply:
Also the generic one can be moved out from the class and shared with other validators

Reply:
Hard to be done. I.e. asset_id and account_id both have std::string type but need to have different validation logics. So template function is not applicable here.

Reply:
Yeah, sorry, for template I mean using a type, e.g string. I mean something like this (note: reason considered a class field):
```cpp
template<typename T>
void validate(T t, std::regex r, std::string &&error) {
  if (not std::regex_match(t, r)) {
    reason.second.push_back(std::move(error));
  }
}
```

And then:
```cpp
void validateAccountId(const interface::types::AccountIdType &account_id) {
  return validate(account_id, R"[a-z]{1,9}\@[a-z]{1,9}");
}
```

Reply:
Thanks